File: daily_python/1014_Best_Sightseeing_Pair.py
import os
import sys
from typing import List

class Solution:
    def maxScoreSightseeingPair(self, values: List[int]) -> int:
        n = len(values)
        ans = float('-inf')
        curr = values[0]
        for i in range(1, n):
            ans = max(ans, curr + values[i] - i)
            curr = max(curr, values[i] + i)
        return ans
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            values = list(map(int, lines[0].split('=')[1].strip()[1:-2].split(',')))
            dataset.append(values)
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for values in dataset:
        results.append(solution.maxScoreSightseeingPair(values))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')

File: daily/2024/1014_Best_Sightseeing_Pair.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int maxScoreSightseeingPair(vector<int>& values) {
        int n = values.size();
        int ans = INT_MIN;
        int curr = values[0];
        for (int i = 1; i < n; i++){
            ans = max(ans, curr + values[i] - i);
            curr = max(curr, values[i] + i);
        }
        return ans;
    }
};

int main(){
    vector<int> values = {8,1,5,2,6};
    // vector<int> values = {1,2};
    Solution S;
    cout << S.maxScoreSightseeingPair(values) << endl;
    return 0;
}